learn/src/main/java/org/navgurukul/learn/courses/repository/LearnRepo.kt

```diff
@@ -0,0 +1,140 @@
+                return courseApi.getCoursesAsync()
+            }
+
+            override suspend fun loadFromDb(): List<Course>? {
+                val data = courseDao.getAllCoursesDirect()
+                data?.forEachIndexed { index, course ->
+                    course.number = (index + 1).toString()
+                }
+                return data
+            }
+        }.asLiveData()
+    }
+
+    fun getCoursesExerciseData(courseId: String): LiveData<List<Exercise>?> {
+        return object : NetworkBoundResource<List<Exercise>, ExerciseResponseContainer>() {
+            override suspend fun saveCallResult(data: ExerciseResponseContainer) {
+                val mappedData = data.data.map {
+                    it.apply { this.courseId = courseId }
+                }.toList()
+                exerciseDao.insertExercise(mappedData)
+            }
+
+            override fun shouldFetch(data: List<Exercise>?): Boolean {
+                //if network avail && shared pref
+                return Util.isConnected(application) && (data == null || data.isEmpty())
+            }
+
+            override suspend fun makeApiCallAsync(): Deferred<ExerciseResponseContainer> {
+                return courseApi.getExercisesAsync(courseId)
+            }
+
+            override suspend fun loadFromDb(): List<Exercise>? {
+                val data = exerciseDao.getAllExercisesForCourseDirect(courseId)
+                parseData(data)
+                return data
+            }
+        }.asLiveData()
+    }
+
+    private fun parseData(data: List<Exercise>) {
+        data.forEachIndexed { index, exercise ->
+            var sequence = (index + 1).toString()
+            if (index + 1 < 10) {
+                sequence = "0" + (index + 1)
+            }
+            exercise.number = sequence
+        }
+    }
+
+    fun getExerciseSlugData(courseId: String, slug: String): LiveData<List<ExerciseSlug>?> {
+        return object : NetworkBoundResource<List<ExerciseSlug>, ExerciseSlug>() {
+            override suspend fun saveCallResult(data: ExerciseSlug) {
+                exerciseSlugDao.insertExerciseSlug(data)
+            }
+
+            override fun shouldFetch(data: List<ExerciseSlug>?): Boolean {
+                //if network avail && shared pref
+                return Util.isConnected(application) && (data == null || data.isEmpty())
+            }
+
+            override suspend fun makeApiCallAsync(): Deferred<ExerciseSlug> {
+                return courseApi.getExercisesSlugAsync(courseId, slug)
+            }
+
+            override suspend fun loadFromDb(): List<ExerciseSlug>? {
+                return exerciseSlugDao.getSlugForExercisesDirect(slug)
+            }
+        }.asLiveData()
+    }
+
+    fun saveCourseExerciseCurrent(currentStudy: CurrentStudy) {
+        Thread(Runnable {
+            currentStudyDao.saveCourseExerciseCurrent(currentStudy)
+        }).start()
```

Comment:
We can make this a suspend function and use dispatcher.io instead of using thread.

Reply:
ViewModel
 ```
viewModelScope.launch {
            learnRepo.saveCourseExerciseCurrent(currentStudy)
  }
```

Repo
```
suspend fun saveCourseExerciseCurrent(currentStudy: CurrentStudy) {
        val currentStudyDao = database.currentStudyDao()
        currentStudyDao.saveCourseExerciseCurrent(currentStudy)
}
```